bundles/org.openhab.binding.solarwatt/src/main/java/org/openhab/binding/solarwatt/internal/handler/EnergyManagerHandler.java

```diff
@@ -0,0 +1,397 @@
+        this.channelTypeProvider = channelTypeProvider;
+    }
+
+    /**
+     * Get services which are provided by this handler.
+     *
+     * Only service discovery is provided by
+     * 
+     * @return collection containing our discovery service
+     */
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return Collections.singleton(SolarwattDevicesDiscoveryService.class);
+    }
+
+    /**
+     * Execute the desired commands.
+     *
+     * Only refresh is supported and relayed to all childs of this thing.
+     *
+     * @param channelUID for which the command is issued
+     * @param command command issued
+     */
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        this.logger.debug("handleCommand called for channel: {}, command: {}", channelUID, command);
+
+        if (command instanceof RefreshType) {
+            this.updateChannels();
+        }
+    }
+
+    /**
+     * Dynnamically updates all known channel states of the energy manager.
+     */
+    public void updateChannels() {
+        Map<String, Device> devices = this.getDevices();
+        if (devices != null) {
+            if (this.energyManagerGuid == null) {
+                try {
+                    this.findEnergyManagerGuid(devices);
+                } catch (SolarwattConnectionException ex) {
+                    this.logger.warn("Failed updating EnergyManager channels: {}", ex.getMessage());
+                }
+            }
+            EnergyManager energyManager = (EnergyManager) devices.get(this.energyManagerGuid);
+
+            if (energyManager != null) {
+                this.calculateUpdates(energyManager);
+
+                energyManager.getStateValues().forEach((stateName, stateValue) -> {
+                    this.updateState(stateName, stateValue);
+                });
+            } else {
+                this.logger.warn("updateChannels failed, missing device EnergyManager {}", this.energyManagerGuid);
+            }
+        }
+    }
+
+    private void calculateUpdates(EnergyManager energyManager) {
+        State timezoneState = energyManager.getState(CHANNEL_IDTIMEZONE.getChannelName());
+        if (timezoneState != null) {
+            this.zoneId = ZoneId.of(timezoneState.toFullString());
+        }
+
+        BigDecimal timestamp = energyManager.getBigDecimalFromChannel(CHANNEL_TIMESTAMP.getChannelName());
+        if (timestamp.compareTo(BigDecimal.ONE) > 0) {
+            energyManager.addState(CHANNEL_DATETIME.getChannelName(),
+                    new DateTimeType(this.getFromMilliTimestamp(timestamp)));
+        }
+    }
+
+    /**
+     * Initial setup of the channels available for this thing.
+     *
+     * @param device which provides the channels
+     */
+    protected void initDeviceChannels(Device device) {
+        this.logger.debug("{}: initDeviceChannels for device {}", this, this.getThing().getUID());
```

Comment:
Can this message be replaced by using the debugger? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging
```
void myFun() {
    logger.trace("Enter myfun"); // DONT, DONT, really DONT do that
    doSomething();
    logger.trace("Leave myfun"); // DONT, DONT, really DONT do that
}
```

Reply:
I removed all trace/debug logging.